fronted.go

```
@@ -48,56 +50,109 @@ type fronted struct {
+		log.Debugf("Creating a new fronted took %v", time.Since(start))
+	}(time.Now())
 
 	f := &fronted{
-		certPool:            pool,
-		masquerades:         loadMasquerades(providersCopy, size),
+		certPool:            atomic.Value{},
+		fronts:              make(sortedFronts, 0),
 		maxAllowedCachedAge: defaultMaxAllowedCachedAge,
 		maxCacheSize:        defaultMaxCacheSize,
 		cacheSaveInterval:   defaultCacheSaveInterval,
 		cacheDirty:          make(chan interface{}, 1),
 		cacheClosed:         make(chan interface{}),
+		providers:           make(map[string]*Provider),
+		clientHelloID:       clientHello,
+		connectingFronts:    newConnectingFronts(4000),
+		stopCh:              make(chan interface{}, 10),
 		defaultProviderID:   defaultProviderID,
-		providers:           providersCopy,
-		clientHelloID:       clientHelloID,
 	}
 
 	if cacheFile != "" {
 		f.initCaching(cacheFile)
 	}
-	f.findWorkingMasquerades(listener)
 
 	return f, nil
 }
 
-func loadMasquerades(initial map[string]*Provider, size int) sortedMasquerades {
-	log.Debugf("Loading candidates for %d providers", len(initial))
+// UpdateConfig sets the domain fronts to use, the trusted root CAs, the fronting providers
+// (such as Akamai, Cloudfront, etc), and the cache file for caching fronts to set up
+// domain fronting.
+func (f *fronted) UpdateConfig(pool *x509.CertPool, providers map[string]*Provider) {
+	// Make copies just to avoid any concurrency issues with access that may be happening on the
+	// caller side.
+	log.Debug("Updating fronted configuration")
+	if len(providers) == 0 {
+		log.Errorf("No providers configured")
+		return
+	}
+	providersCopy := copyProviders(providers)
+	f.frontedMu.Lock()
+	defer f.frontedMu.Unlock()
+	f.addProviders(providersCopy)
+	f.addFronts(loadFronts(providersCopy))
+
+	f.certPool.Store(pool)
+
+	// The goroutine for finding working fronts runs forever, so only start it once.
+	f.crawlOnce.Do(func() {
+		go f.findWorkingFronts()
+	})
+}
```

This just occurred to me: we should return an error here if `fronted` has been stopped or track whether `findWorkingFronts` is running or not and restart it if we receive a new config after it's been stopped. Using `sync.Once` could be problematic. If a goroutine doesn't check `fronted` hasn't been Closed, it could call `UpdateConfig` and go on happy assuming `findWorkingFronts` is running, which could be difficult to debug.

---

Hmnn interesting. The only time `Close` is _intended_ to be called is when the app is closed, but I suppose safeguarding against random closing is a good idea.

---

I honestly think we should trust callers to not randomly stop fronted